### PR TITLE
Use 'HttpMessageHandler' instead of 'HttpClientHandler' to support NSUrlSessionHandler on iOS

### LINKED ISSUE #2502: SentryOptions should use 'HttpMessageHandler' instead of 'HttpClientHandler' to support NSUrlSessionHandler on iOS
### Problem Statement

The SentryOptions should accept the more general HttpMessageHandler instead of HttpClientHandler to provide to ability to use a custom-configured handler for the HttpClient.

Background:
It is recommended to use NSUrlSessionHandler on iOS (TLS 1.2, performance), but NSUrlSessionHandler only implements HttpMessageHandler.

Since NSUrlSessionHandler inherits from HttpMessageHandler and DefaultSentryHttpClientFactory also mainly uses this class to create the HttpClient, changes are going to be minor.

During creation of the HttpClient the DefaultSentryHttpClientFactory can take care of if it is a HttpClientHandler, a HttpMessageHandler or a NSUrlSessionHandler (which also accepts proxy and decompression properties.

### Solution Brainstorm

Going to file a PR.

### PR DESCRIPTION
**Pull request by AlexSchuetz**
fixing #2502 

Breaking change:
 - SentryOptions.HttpClientHandlerFactory moved to SentryOptions.HttpMessageHandlerFactory

### CODE REVIEW

**Comment by bitsandfoxes** [comment]
That's a great catch, thanks for contributing!
We'd like to avoid introducing breaking changes outside of mayor releases and the next one is still a bit out.
I'm thinking, could this be added as an additional option and we mark the old one as `obsolete`?

**Comment by AlexSchuetz** [comment]
Of course. That's a good idea. I will adjust it shortly. 👍

**Comment by jamescrosswell** [comment]
Thanks heaps @AlexSchuetz, the commit all looks great!

There are a couple of things that need doing to get things through our CI servers though.

### Changelog

Could you add the following to the CHANGELOG.md in the Fixes section?

```
- Fixes #2502 SentryOptions should use 'HttpMessageHandler' instead of 'HttpClientHandler' to support NSUrlSessionHandler on iOS  ([#2503](https://github.com/getsentry/sentry-dotnet/pull/2503))
```

### Update verify expectations
We're using [Verify](https://github.com/VerifyTests/Verify) to make sure that any changes to the API surface area are deliberate. That means when the API does change, we need to update the relevant verified.txt files. The easiest way to create updated versions of those is to run `./build.sh` or `build.cmd` (depending on your development platform) from the repository root folder. So if you're developing on MacOS, for example, you'd run `./build.sh`. Once you've done that, I think you should see changes to commit for the following files:
- test/Sentry.Tests/ApiApprovalTests.Run.Net4_8.verified.txt
- test/Sentry.Tests/ApiApprovalTests.Run.Core3_1.verified.txt
- test/Sentry.Tests/ApiApprovalTests.Run.DotNet6_0.verified.txt
- test/Sentry.Tests/ApiApprovalTests.Run.DotNet7_0.verified.txt

You can add those to a new commit with a description like "updated verify files" or something like that and push.

**_Note:_** You can ignore any errors the build shows relating to profiling... that's a long story and unrelated to this PR.

### Failed Checks
There are a [couple of other checks that are failing](https://github.com/getsentry/sentry-dotnet/actions/runs/5623842449/job/15273647829?pr=2503). I'm not actually sure what's going on there. @bitsandfoxes do you have any context for that?

**Comment by AlexSchuetz** [comment]
I did the required commit and also had a look at the error. NSUrlSessionHandler does not support a proxy, but I want to include the code for setting it anyway, because if it ever does noone will remember adjusting it here.
Thus I implemented it and placed it inside of a try-catch. So nothing should go wrong. I now adjusted my implementation and check NSUrlSessionHandler.SupportsProxy. Hopefully this works.

**Comment by AlexSchuetz** [comment]
I investigated a bit more. It is a code analysis error. Since NSUrlSessionHandler.Proxy is not supported.
I added a pragma for this one line to disable this issue, since I want to support it as soon as NSUrlSesseionHandler does.

If you think I should remove the code in question and just print out that a proxy for NSUrlSessionHandler is not supported, that would be fine as well. I don't know if the Proxy-property will ever be implemented. I think iOS and MAC require a proxy to be defined elsewhere.

File: src/Sentry/Internal/Http/DefaultSentryHttpClientFactory.cs [hunk 1]
```diff
@@ -20,25 +20,77 @@ public HttpClient Create(SentryOptions options)
...
+                        options.LogInfo("Using Proxy: {0}", options.HttpProxy);
+                    }
+                    catch (PlatformNotSupportedException)
+                    {
+                        supportsProxy = false;
+                    }
+                }
+                if (!supportsProxy)
+                {
+                    options.LogDebug("No proxy supported by NSUrlSessionHandler.");
```
Comment (bruno-garcia) [inline src/Sentry/Internal/Http/DefaultSentryHttpClientFactory.cs:69]: User was explicit in setting a proxy, lets warn them it didn't work:

```suggestion
                    options.LogWarning("No proxy supported by NSUrlSessionHandler.");
```

File: src/Sentry/SentryOptions.cs [hunk 2]
```diff
@@ -542,8 +542,19 @@ public int MaxCacheItems
 
     /// <summary>
     /// Creates the inner most <see cref="HttpClientHandler"/>.
+    /// Deprecated in favor of <see cref="CreateHttpMessageHandler"/>.
     /// </summary>
-    public Func<HttpClientHandler>? CreateHttpClientHandler { get; set; }
+    [Obsolete("Use CreateHttpMessageHandler instead")]
+    public Func<HttpClientHandler>? CreateHttpClientHandler
+    {
+        get => CreateHttpMessageHandler as Func<HttpClientHandler>;
```
Comment (bruno-garcia) [inline src/Sentry/SentryOptions.cs:550]: Does this typecast work? If not, it'll just return `null`

File: src/Sentry/Internal/Http/DefaultSentryHttpClientFactory.cs [hunk 3]
```diff
@@ -20,25 +20,77 @@ public HttpClient Create(SentryOptions options)
...
+                    compressionSupported = true;
+                }
+            }
+            catch (PlatformNotSupportedException)
+            {
+                compressionSupported = false;
+            }
+            if (!compressionSupported)
+            {
+                options.LogDebug("No response compression supported by NSUrlSessionHandler.");
```
Comment (bruno-garcia) [inline src/Sentry/Internal/Http/DefaultSentryHttpClientFactory.cs:90]: Ideally we'd be aware of the intention of the user, as in, did they set a non-default value to `DecompressionMethods`? If so we might want to `Warn` that things didn't go as they expected instead. But at least LogInfo is a better choice then Debug here:

```suggestion
                options.LogInfo("No response compression supported by NSUrlSessionHandler.");
```

**Review by bruno-garcia — APPROVED** [review]
Left some minor suggestions but generally lgtm

File: src/Sentry/SentryOptions.cs (hunk 2 above)
Comment (AlexSchuetz) [reply src/Sentry/SentryOptions.cs]: Unfortunately you are right. This doesn't work. I introduced a private field to hold the `Func<HttpClientHandler>?`. So when the setter is used, the getter will work. if the new `CreateHttpMessageHandler` is used instead `CreateHttpClientHandler`will return null.

**Comment by AlexSchuetz** [comment]
@bruno-garcia I followed your suggestions. Am I supposed to resolve those conversations now?

**Comment by bitsandfoxes** [comment]
> Am I supposed to resolve those conversations now?

Yes. Everything that has been addressed can be marked as resolved.

File: src/Sentry/SentryOptions.cs [hunk 4]
```diff
@@ -540,10 +540,30 @@ public int MaxCacheItems
...
+
     /// <summary>
     /// Creates the inner most <see cref="HttpClientHandler"/>.
+    /// Deprecated in favor of <see cref="CreateHttpMessageHandler"/>.
+    /// </summary>
+    [Obsolete("Use CreateHttpMessageHandler instead")]
+    public Func<HttpClientHandler>? CreateHttpClientHandler
+    {
+        get => _createClientHandler;
+        set =>
```
Comment (bitsandfoxes) [inline src/Sentry/SentryOptions.cs:556]:
```suggestion
        set
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -11,6 +11,7 @@
 - Fixed baggage propagation when an exception is thrown from middleware ([#2487](https://github.com/getsentry/sentry-dotnet/pull/2487))
 - Fix Durable Functions preventing orchestrators from completing ([#2491](https://github.com/getsentry/sentry-dotnet/pull/2491))
 - Re-enable HubTests.FlushOnDispose_SendsEnvelope ([#2492](https://github.com/getsentry/sentry-dotnet/pull/2492))
+- Fixes #2502 SentryOptions should use 'HttpMessageHandler' instead of 'HttpClientHandler' to support NSUrlSessionHandler on iOS  ([#2503](https://github.com/getsentry/sentry-dotnet/pull/2503))
```
Comment (bitsandfoxes) [inline CHANGELOG.md:14]:
```suggestion
- Introduced `HttpMessageHandler` in favor of the now deprecated `HttpClientHandler` on the options. This allows the SDK to support NSUrlSessionHandler on iOS ([#2503](https://github.com/getsentry/sentry-dotnet/pull/2503))
```

**Review by bitsandfoxes — APPROVED** [review]
This looks good to me! Thanks again for contributing!

**Comment by AlexSchuetz** [comment]
@bitsandfoxes @bruno-garcia  You are welcome. Thank's for your support and review.

**Pull request merged**